### ISSUE TITLE
git is broken after rename of repo root with submodules (only after git clone --recursive)

### ISSUE THREAD

**Issue by AArnott** [git-for-windows/git#756]
- [X] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen)
  or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue
  matching what I'm seeing
### Setup
- Which version of Git for Windows are you using? 32-bit or 64-bit? Include the
  output of `git version` as well.

```
$ git --version
git version 2.8.2.windows.1
#64-bit
```
- Which version of Windows are you running? 32-bit or 64-bit?

64-bit Windows 10
- What options did you set as part of the installation? Or did you choose the
  defaults?

defaults
- Any other interesting things about your environment that might be related
  to the issue you're seeing?

Known to repro on multiple Windows machines _and linux_
### Details
- Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Powershell
- What commands did you run to trigger this issue? If you can provide a
  [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
  this will help us understand the issue.

```
C:\Repos>git clone --recursive https://mseng.visualstudio.com/DefaultCollection/VSIDEProj/_git/VSIDEProj.Threading Threading
Cloning into 'Threading'...
remote:
remote:                    vSTs
remote:                  vSTSVSTSv
remote:                vSTSVSTSVST
remote: VSTS         vSTSVSTSVSTSV
remote: VSTSVS     vSTSVSTSV STSVS
remote: VSTSVSTSvsTSVSTSVS   TSVST
remote: VS  tSVSTSVSTSv      STSVS
remote: VS   tSVSTSVST       SVSTS
remote: VS tSVSTSVSTSVSts    VSTSV
remote: VSTSVST    SVSTSVSTs VSTSV
remote: VSTSv        STSVSTSVSTSVS
remote:                VSTSVSTSVST
remote:                  VSTSVSTs
remote:                    VSTs    (TM)
remote:
remote:  Microsoft (R) Visual Studio (R) Team Services
remote:
Receiving objects: 100% (10345/10345), 5.35 MiB | 9.33 MiB/s, done.
Resolving deltas: 100% (5834/5834), done.
Checking connectivity... done.
Submodule 'Razzle' (https://mseng.visualstudio.com/DefaultCollection/VSIDEProj/_git/VSIDEProj.RazzleDependencies) registered for path 'Razzle'
Cloning into 'Razzle'...
remote:
remote:                    vSTs
remote:                  vSTSVSTSv
remote:                vSTSVSTSVST
remote: VSTS         vSTSVSTSVSTSV
remote: VSTSVS     vSTSVSTSV STSVS
remote: VSTSVSTSvsTSVSTSVS   TSVST
remote: VS  tSVSTSVSTSv      STSVS
remote: VS   tSVSTSVST       SVSTS
remote: VS tSVSTSVSTSVSts    VSTSV
remote: VSTSVST    SVSTSVSTs VSTSV
remote: VSTSv        STSVSTSVSTSVS
remote:                VSTSVSTSVST
remote:                  VSTSVSTs
remote:                    VSTs    (TM)
remote:
remote:  Microsoft (R) Visual Studio (R) Team Services
remote:
Receiving objects: 100% (1506/1506), 114.50 MiB | 16.75 MiB/s, done.
Resolving deltas: 100% (580/580), done.
Checking connectivity... done.
Submodule path 'Razzle': checked out '02a8f0566140a0152e4d2b5fb861163d5d73e29d'

C:\Repos>cd Threading

C:\Repos\Threading>git status
On branch v14
Your branch is up-to-date with 'origin/v14'.
nothing to commit, working directory clean


[2]
C:\Repos>git clone --recursive https://mseng.visualstudio.com/DefaultCollection/VSIDEProj/_git/VSIDEProj.Threading
Cloning into 'VSIDEProj.Threading'...
remote:
remote:                    vSTs
remote:                  vSTSVSTSv
remote:                vSTSVSTSVST
remote: VSTS         vSTSVSTSVSTSV
remote: VSTSVS     vSTSVSTSV STSVS
remote: VSTSVSTSvsTSVSTSVS   TSVST
remote: VS  tSVSTSVSTSv      STSVS
remote: VS   tSVSTSVST       SVSTS
remote: VS tSVSTSVSTSVSts    VSTSV
remote: VSTSVST    SVSTSVSTs VSTSV
remote: VSTSv        STSVSTSVSTSVS
remote:                VSTSVSTSVST
remote:                  VSTSVSTs
remote:                    VSTs    (TM)
remote:
remote:  Microsoft (R) Visual Studio (R) Team Services
remote:
Receiving objects: 100% (10345/10345), 5.35 MiB | 2.34 MiB/s, done.
Resolving deltas: 100% (5834/5834), done.
Checking connectivity... done.
Submodule 'Razzle' (https://mseng.visualstudio.com/DefaultCollection/VSIDEProj/_git/VSIDEProj.RazzleDependencies) registered for path 'Razzle'
Cloning into 'Razzle'...
remote:
remote:                    vSTs
remote:                  vSTSVSTSv
remote:                vSTSVSTSVST
remote: VSTS         vSTSVSTSVSTSV
remote: VSTSVS     vSTSVSTSV STSVS
remote: VSTSVSTSvsTSVSTSVS   TSVST
remote: VS  tSVSTSVSTSv      STSVS
remote: VS   tSVSTSVST       SVSTS
remote: VS tSVSTSVSTSVSts    VSTSV
remote: VSTSVST    SVSTSVSTs VSTSV
remote: VSTSv        STSVSTSVSTSVS
remote:                VSTSVSTSVST
remote:                  VSTSVSTs
remote:                    VSTs    (TM)
remote:
remote:  Microsoft (R) Visual Studio (R) Team Services
remote:
Receiving objects: 100% (1506/1506), 114.50 MiB | 8.46 MiB/s, done.
Resolving deltas: 100% (580/580), done.
Checking connectivity... done.
Submodule path 'Razzle': checked out '02a8f0566140a0152e4d2b5fb861163d5d73e29d'

C:\Repos>cd VSIDEProj.Threading

C:\Repos\VSIDEProj.Threading>git status
On branch v14
Your branch is up-to-date with 'origin/v14'.
nothing to commit, working directory clean

```
- What did you expect to occur after running these commands?

Git status and other commands continue to work properly.
- What actually happened instead?

Git started malfunctioning in very basic commands after the rename.

**Comment by AArnott** [comment]
Ryan Roden-Corrent reports:

> The recursive clone sets up an absolute path for gitdir in the .git file of each submodule.
When you move the repo these all get broken.
I'm surprised I haven't heard more noise about this, the closest I can find is:
http://stackoverflow.com/questions/9878860/how-can-i-rename-a-git-repository-with-submodules#9885513

> If it is a bug, its in core git, not git for windows (just tried this on linux, same result).
`git submodule update --init --recursive` has the same result as well.

J Wyman shared:

> From the git-submodule script:

```
# The function takes at most 2 arguments. The first argument is the
# URL that navigates to the submodule origin repo. When relative, this URL
# is relative to the superproject origin URL repo. The second up_path
# argument, if specified, is the relative path that navigates
# from the submodule working tree to the superproject working tree.
#
# The output of the function is the origin URL of the submodule.
#
# The output will either be an absolute URL or filesystem path (if the
# superproject origin URL is an absolute URL or filesystem path,
# respectively) or a relative file system path (if the superproject
# origin URL is a relative file system path).
#
# When the output is a relative file system path, the path is either
# relative to the submodule working tree, if up_path is specified, or to
# the superproject working tree otherwise.
```

> Looks it would be one of two options:

> 1)	Change the git-checkout built-in to issue a relative path to the sub-module script.
> 2)	Change the sub-module script to always use a relative path if it possibly can.

**Comment by dscho** [comment]
It looks as if the bug will be fixed in v2.9.0 (due later this month). From upstream's tentative release notes:

>  * A partial rewrite of "git submodule" in the 2.7 timeframe changed the way the gitdir: pointer in the submodules point at the real repository location to use absolute paths by accident.  This has been corrected.

Of course this won't fix the absolute paths that have been written already; Unfortunately you will have to fix them manually.

**Comment by AArnott** [comment]
Sounds reasonable. Thanks.
